lib/functions_framework/function.rb

```diff
@@ -164,6 +181,66 @@ def call *args, globals: nil, logger: nil
       callable.call(*args)
     end
 
+    ##
+    # Wraps the provided callable or block by returning a new anonymous
+    # class implementing callable that adds request deserialization and response
+    # serialization.
+    # @private
+    #
+    # @param callable [Class,#call] A callable object or class.
+    # @param request_class [Class] The class of the ruqest if the function is 'typed'
+    # @param block [Proc] The function code as a block.
+    #
+    def wrap_typed callable, request_class: nil, &block
+      if block_given?
+        callable = ::Class.new Callable do
+          define_method :call, &block
+        end
+      end
+      if request_class
+        unless request_class.respond_to? :decode_json
+          raise ::ArgumentError, "Type does not implement 'decode_json' class method"
+        end
+        unless request_class.method_defined? :to_json
+          raise ::ArgumentError, "Type does not implement 'to_json' instance method"
+        end
+      end
+      if callable.is_a? ::Class
+        return make_typed_callable nil, callable_class: callable, request_class: request_class
+      end
+      make_typed_callable callable, request_class: request_class
+    end
+
+    ##
+    # Accepts a callable and wraps it with request decoding and response encoding
+    # @private
+    #
+    def make_typed_callable callable, callable_class: nil, request_class: nil
+      ::Class.new Callable do
+        define_method :call do |http_req|
+          begin
+            req = if request_class
+                    request_class.decode_json http_req.body.read.to_s
+                  else
+                    JSON.parse http_req.body.read.to_s
+                  end
+          rescue JSON::ParserError
+            return [400, "text/plain", "Bad Request"]
+          end
+
+          to_call = if callable_class
+                      callable_class.new globals: @__globals, logger: @__logger
+                    else
+                      callable
+                    end
+          res = to_call.call req
+
+          return [200, "application/json", res.to_json] if request_class
```

Comment:
for my understanding, what's the difference between to_json and JSON.generate?

Reply:
Is there encoding in this logic? not sure if that was made clear in here

Reply:
This I had to look into a bit, I didn't realize this but it seems they are equivalent (you can even call nil.to_json). I've removed the JOSN.generate line in favor of just using res.to_json.